File: javascript/exercises/practice/react/react.js
/**
 * Reactive Programming System.
 * 
 * This module implements a spreadsheet-like reactive system using a Directed Acyclic Graph (DAG)
 * of cells. It consists of:
 * - `InputCell`: Value sources that can be manually updated.
 * - `ComputeCell`: Derived cells whose values depend on other cells and are updated automatically.
 * - `CallbackCell`: Observers registered on compute cells that record state changes.
 * 
 * Architectural Highlights:
 * 1. Topological Sorting: Updates propagate from modified InputCells to dependent ComputeCells in
 *    topological order. This guarantees each cell is recalculated exactly once and always has the latest
 *    values from its inputs, preventing glitching or inconsistent intermediate states.
 * 2. Stable State Callbacks: Callbacks are only invoked after the entire network has settled into a
 *    stable state. A callback is triggered only if the cell's final value differs from its value at the
 *    start of the transaction (setValue invocation).
 */

export class InputCell {
  /**
   * Creates a new input cell holding a primitive value.
   * @param {*} value - Initial value of the cell.
   */
  constructor(value) {
    this.value = value;
    /** @type {Set<ComputeCell>} Direct descendants in the reactive graph */
    this.dependents = new Set();
  }

  /**
   * Sets a new value for the input cell and triggers propagation through the dependency graph.
   * Updates only run if the new value is different from the current value (strict equality).
   * @param {*} value - The new value to assign.
   */
  setValue(value) {
    if (this.value === value) return;

    // 1. Traverse and discover all transitively dependent ComputeCells.
    // We must capture their values before any recomputations occur to accurately
    // determine if their final values in the new stable state have changed.
    const dependentComputeCells = [];
    const visited = new Set();
    
    const findComputeCells = (cell) => {
      for (const dep of cell.dependents) {
        if (!visited.has(dep)) {
          visited.add(dep);
          dependentComputeCells.push(dep);
          findComputeCells(dep);
        }
      }
    };
    findComputeCells(this);

    // Snapshot original stable values of the dependent compute cells.
    const originalValues = new Map();
    for (const cell of dependentComputeCells) {
      originalValues.set(cell, cell.value);
    }

    // Set the new value on this input cell.
    this.value = value;

    // 2. Perform a topological sort of all dependent compute cells.
    // This ensures that for any cell C depending on cells A and B, C is recomputed
    // only after both A and B are fully recomputed and stable.
    const topoOrder = [];
    const topoVisited = new Set();
    
    const topoVisit = (cell) => {
      if (topoVisited.has(cell)) return;
      topoVisited.add(cell);
      for (const dep of cell.dependents) {
        if (visited.has(dep)) {
          topoVisit(dep);
        }
      }
      topoOrder.push(cell);
    };

    for (const cell of dependentComputeCells) {
      topoVisit(cell);
    }
    // Reversing the post-order depth-first traversal yields a valid topological sort.
    topoOrder.reverse();

    // 3. Recompute each ComputeCell in topological order.
    for (const cell of topoOrder) {
      cell.value = cell.calculateValue();
    }

    // 4. Trigger callbacks for those compute cells whose values changed
    // from their original stable value to their final stable value.
    for (const cell of topoOrder) {
      const origVal = originalValues.get(cell);
      if (origVal !== cell.value) {
        // Create a shallow copy of the callback list to avoid mutation issues
        // if callbacks register or deregister other callbacks during invocation.
        for (const cb of Array.from(cell.callbacks)) {
          const val = cb.fn(cell);
          cb.values.push(val);
        }
      }
    }
  }
}

export class ComputeCell {
  /**
   * Creates a computed cell that automatically updates its value based on dependencies.
   * @param {Array<InputCell|ComputeCell>} inputCells - Upstream dependency cells.
   * @param {function(Array<InputCell|ComputeCell>): *} fn - Computation function.
   */
  constructor(inputCells, fn) {
    this.inputs = inputCells;
    this.fn = fn;
    /** @type {Set<ComputeCell>} Direct descendants in the reactive graph */
    this.dependents = new Set();
    /** @type {Set<CallbackCell>} Registered change notification observers */
    this.callbacks = new Set();
    
    // Register this cell as a dependent of all its input dependency cells.
    for (const input of this.inputs) {
      input.dependents.add(this);
    }
    
    // Compute the initial value.
    this.value = this.calculateValue();
  }

  /**
   * Computes the current value of the cell using the user-provided function.
   * @returns {*} Computed value.
   */
  calculateValue() {
    return this.fn(this.inputs);
  }

  /**
   * Registers a change notification callback cell.
   * @param {CallbackCell} cb - Callback cell to add.
   */
  addCallback(cb) {
    this.callbacks.add(cb);
  }

  /**
   * Removes a registered change notification callback cell.
   * @param {CallbackCell} cb - Callback cell to remove.
   */
  removeCallback(cb) {
    this.callbacks.delete(cb);
  }
}

export class CallbackCell {
  /**
   * Creates an observer that records state change notifications of a ComputeCell.
   * @param {function(ComputeCell): *} fn - Function to run on change.
   */
  constructor(fn) {
    this.fn = fn;
    /** @type {Array<*>} List of recorded return values from the callback function */
    this.values = [];
  }
}

File: go/exercises/practice/react/react.go
package react

import (
	"container/heap"
)

// reactiveCell defines the internal interface for cells participating in dependency tracking and propagation.
type reactiveCell interface {
	Cell
	getLevel() int
	addConsumer(c *computeCell)
}

// reactor manages linked cells.
type reactor struct{}

// inputCell holds a settable value and propagates changes downstream.
type inputCell struct {
	r         *reactor
	val       int
	consumers []*computeCell
}

// callbackEntry stores a registered callback and its unique identifier for cancellation.
type callbackEntry struct {
	id int
	fn func(int)
}

// computeCell computes its value dynamically from upstream dependencies and triggers callbacks on stable changes.
type computeCell struct {
	r              *reactor
	val            int
	stableVal      int
	level          int
	computeFn      func() int
	consumers      []*computeCell
	callbacks      []callbackEntry
	nextCallbackID int
	inQueue        bool
}

// canceler removes a registered callback from its corresponding compute cell.
type canceler struct {
	cell *computeCell
	id   int
}

// Value returns the current value of the input cell.
func (c *inputCell) Value() int {
	return c.val
}

// getLevel returns the topological level (0 for inputs).
func (c *inputCell) getLevel() int {
	return 0
}

// addConsumer registers a downstream compute cell that depends on this input cell.
func (c *inputCell) addConsumer(consumer *computeCell) {
	c.consumers = append(c.consumers, consumer)
}

// SetValue sets the value of the input cell and triggers topological update propagation.
func (c *inputCell) SetValue(value int) {
	if c.val == value {
		return
	}
	c.val = value
	c.r.propagate(c.consumers)
}

// Value returns the current value of the compute cell.
func (c *computeCell) Value() int {
	return c.val
}

// getLevel returns the topological level of the compute cell.
func (c *computeCell) getLevel() int {
	return c.level
}

// addConsumer registers a downstream compute cell that depends on this compute cell.
func (c *computeCell) addConsumer(consumer *computeCell) {
	c.consumers = append(c.consumers, consumer)
}

// AddCallback registers a callback to be called when the cell's stable value changes.
func (c *computeCell) AddCallback(callback func(int)) Canceler {
	id := c.nextCallbackID
	c.nextCallbackID++
	c.callbacks = append(c.callbacks, callbackEntry{id: id, fn: callback})
	return &canceler{
		cell: c,
		id:   id,
	}
}

// removeCallback removes a callback by its identifier.
func (c *computeCell) removeCallback(id int) {
	for i, cb := range c.callbacks {
		if cb.id == id {
			c.callbacks = append(c.callbacks[:i], c.callbacks[i+1:]...)
			return
		}
	}
}

// hasCallback checks if a callback with the given identifier is currently active.
func (c *computeCell) hasCallback(id int) bool {
	for _, cb := range c.callbacks {
		if cb.id == id {
			return true
		}
	}
	return false
}

// Cancel removes the registered callback from the compute cell.
func (c *canceler) Cancel() {
	if c.cell == nil {
		return
	}
	c.cell.removeCallback(c.id)
	c.cell = nil
}

// cellHeap implements container/heap.Interface for compute cells ordered by topological level.
type cellHeap []*computeCell

func (h cellHeap) Len() int           { return len(h) }
func (h cellHeap) Less(i, j int) bool { return h[i].level < h[j].level }
func (h cellHeap) Swap(i, j int)      { h[i], h[j] = h[j], h[i] }
func (h *cellHeap) Push(x any)        { *h = append(*h, x.(*computeCell)) }
func (h *cellHeap) Pop() any {
	old := *h
	n := len(old)
	x := old[n-1]
	*h = old[0 : n-1]
	return x
}

// propagate computes updated values in topological order and fires callbacks once a stable state is reached.
func (r *reactor) propagate(initialConsumers []*computeCell) {
	if len(initialConsumers) == 0 {
		return
	}

	h := make(cellHeap, 0, len(initialConsumers))
	for _, c := range initialConsumers {
		if !c.inQueue {
			c.inQueue = true
			heap.Push(&h, c)
		}
	}

	var changedComputeCells []*computeCell

	for h.Len() > 0 {
		c := heap.Pop(&h).(*computeCell)
		c.inQueue = false

		newVal := c.computeFn()
		if newVal != c.val {
			c.val = newVal
			changedComputeCells = append(changedComputeCells, c)
			for _, consumer := range c.consumers {
				if !consumer.inQueue {
					consumer.inQueue = true
					heap.Push(&h, consumer)
				}
			}
		}
	}

	for _, c := range changedComputeCells {
		if c.val != c.stableVal {
			c.stableVal = c.val
			cbs := make([]callbackEntry, len(c.callbacks))
			copy(cbs, c.callbacks)
			for _, cb := range cbs {
				if c.hasCallback(cb.id) {
					cb.fn(c.val)
				}
			}
		}
	}
}

// New creates and returns a new Reactor instance.
func New() Reactor {
	return &reactor{}
}

// CreateInput creates an input cell linked into the reactor with the given initial value.
func (r *reactor) CreateInput(initial int) InputCell {
	return &inputCell{
		r:   r,
		val: initial,
	}
}

// CreateCompute1 creates a compute cell which computes its value based on one other cell.
func (r *reactor) CreateCompute1(dep Cell, compute func(int) int) ComputeCell {
	rc := dep.(reactiveCell)
	level := rc.getLevel() + 1
	initialVal := compute(dep.Value())

	c := &computeCell{
		r:         r,
		val:       initialVal,
		stableVal: initialVal,
		level:     level,
		computeFn: func() int {
			return compute(dep.Value())
		},
	}
	rc.addConsumer(c)
	return c
}

// CreateCompute2 creates a compute cell which computes its value based on two other cells.
func (r *reactor) CreateCompute2(dep1, dep2 Cell, compute func(int, int) int) ComputeCell {
	rc1 := dep1.(reactiveCell)
	rc2 := dep2.(reactiveCell)
	level := rc1.getLevel()
	if rc2.getLevel() > level {
		level = rc2.getLevel()
	}
	level++

	initialVal := compute(dep1.Value(), dep2.Value())

	c := &computeCell{
		r:         r,
		val:       initialVal,
		stableVal: initialVal,
		level:     level,
		computeFn: func() int {
			return compute(dep1.Value(), dep2.Value())
		},
	}
	rc1.addConsumer(c)
	rc2.addConsumer(c)
	return c
}
